helix-term/src/grammars.rs

```diff
@@ -0,0 +1,290 @@
+            job(grammar);
+
+            // report progress
+            tx.send(1).unwrap();
+        });
+    }
+    pool.join();
+
+    assert_eq!(rx.try_iter().sum::<usize>(), n_jobs);
+}
+
+pub fn fetch_grammar(grammar: GrammarConfiguration) {
+    if let GrammarSource::Git { remote, revision } = grammar.source {
+        let grammar_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("../runtime/grammars/sources")
+            .join(grammar.grammar_id.clone());
+
+        fs::create_dir_all(grammar_dir.clone()).expect("Could not create grammar directory");
+
+        // create the grammar dir contains a git directory
+        if !grammar_dir.join(".git").is_dir() {
+            Command::new("git")
+                .args(["init"])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Could not execute 'git'");
+        }
+
+        // ensure the remote matches the configured remote
+        if get_repository_info(&grammar_dir, vec!["remote", "get-url", REMOTE_NAME])
+            != Some(remote.clone())
+        {
+            set_remote(&grammar_dir, &remote);
+        }
+
+        // ensure the revision matches the configured revision
+        if get_repository_info(&grammar_dir, vec!["rev-parse", "HEAD"]) != Some(revision.clone()) {
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            Command::new("git")
+                .args(["fetch", REMOTE_NAME, &revision])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Failed to execute 'git'");
+
+            Command::new("git")
+                .args(["checkout", &revision])
+                .current_dir(grammar_dir)
+                .output()
+                .expect("Failed to execute 'git'");
+
+            println!(
+                "Grammar '{}' checked out at '{}'.",
+                grammar.grammar_id, revision
+            );
+        } else {
+            println!("Grammar '{}' is already up to date.", grammar.grammar_id);
+        }
+    };
+}
+
+// Sets the remote for a repository to the given URL, creating the remote if
+// it does not yet exist.
+fn set_remote(repository: &Path, remote_url: &str) {
+    if !Command::new("git")
+        .args(["remote", "set-url", REMOTE_NAME, remote_url])
+        .current_dir(repository)
+        .output()
+        .expect("Failed to execute 'git'")
+        .status
+        .success()
+        && !Command::new("git")
+            .args(["remote", "add", REMOTE_NAME, remote_url])
+            .current_dir(repository)
+            .output()
+            .expect("Failed to execute 'git'")
+            .status
+            .success()
+    {
+        eprintln!("Failed to set remote '{}'", remote_url);
+    }
+}
+
+fn get_repository_info(repository: &Path, args: Vec<&str>) -> Option<String> {
```

Comment:
This is a bit of an odd function. It can run an arbitrary git command, but it assumes that the output will be a single line. I'm not sure this is better than just doing `git set-url` and `git rev-parse` separately, especially so if they fail, you know which one.

Reply:
Hmm yeah originally I had them separated but the code was basically duplicated between them. They're so short that probably doesn't matter - I'll separate them again